### PR TITLE
[batch] Setup database infrastructure for job groups

### PR DESCRIPTION
**Pull request by jigold**
The conceptual change here is we want to parameterize all batch related tables to have a new job group ID that I've set to **0** for the root job group. We need to make sure all future inserts / updates into the batches table are propagated to the new job groups table. When we create a batch now, we also create the corresponding entries into the job groups and job group parents tables.

I chose the root job group to be 0 as I think conceptually, the client should start numbering job groups at 1 and not know there is a hidden root job group being created under the hood. I'm not wedded to this.

I tried to check for all the indices that would be needed in my prototype. It's possible I missed one or two, but it's not a big deal to add it later.

I don't think we need to test this on a populated database (dev deploy main, submit jobs, then run the migration), but let me know if you think that would be helpful.

### CODE REVIEW

File: batch/batch/front_end/front_end.py
Comment (daniel-goldstein) [inline batch/batch/front_end/front_end.py:87]: I'm all for making `ROOT_JOB_GROUP_ID` a constant, but let's move it out of globals because to me that gives the impression that it is configurable or might change in the future. IIUC it is basically never a good idea to change this constant.

File: batch/sql/setup-job-groups.sql
Comment (daniel-goldstein) [inline batch/sql/setup-job-groups.sql:80]: For clarification, are we going to need this part of the trigger forever to ensure that the duplicated state between job groups and batches is updated in unison?

Comment (daniel-goldstein) [reply batch/sql/setup-job-groups.sql:80]: Do we want to allow things like the `user` to change?

File: batch/sql/estimated-current.sql
Comment (daniel-goldstein) [inline batch/sql/estimated-current.sql:222]: nit: I'd prefer to use the full name in the indexes ( `parent_id` )

File: batch/sql/estimated-current.sql
Comment (daniel-goldstein) [inline batch/sql/estimated-current.sql:1436]: To confirm my understanding, cancelling the batch means a row that is `(batch_id, NULL)` and cancelling a job group means a row that is `(batch_id, job_group_id)`?

**Review by daniel-goldstein — CHANGES_REQUESTED** [review]
Sorry for the delay on this, I wanted to make sure I could devote full attention to it. It looks good, I just have some clarifying questions.

File: batch/sql/estimated-current.sql
Comment (jigold) [reply batch/sql/estimated-current.sql:1436]: No. There's an implicit default value here of job_group_id = 0 (`INT DEFAULT 0`). I can explicitly add the job_group_id being 0 here for clarity.

File: batch/sql/setup-job-groups.sql
Comment (jigold) [reply batch/sql/setup-job-groups.sql:80]: No. The `user` is for an index that is needed. It shouldn't change between job groups in the same batch, but I wanted to be complete here that any field updates are reflected in the job groups table.

As for needing the trigger forever, I think we have independent insert and update commands in the batch front end / driver that keep the state in sync and not in the trigger. At some point, I want most of these fields to be essentially deprecated on the batches table. Thoughts??

File: batch/batch/front_end/front_end.py
Comment (jigold) [reply batch/batch/front_end/front_end.py:87]: Yep, this constant can never change once we choose it.

File: batch/sql/setup-job-groups.sql
Comment (daniel-goldstein) [reply batch/sql/setup-job-groups.sql:80]: ah, ya I appreciate wanting to be complete here but as a reader I interpret this less as being complete and more as "accounting for all of the fields that could change". Can we instead error if the new value differs from the old one?

**Comment by jigold** [comment]
I'm putting this on hold until I get clarity on #13513

**Comment by jigold** [comment]
@danking I'd like you to look over the new SQL in MJC as well.

**Comment by jigold** [comment]
This is now passing tests in gcp. I didn't retry it in Azure, but I presume it would pass as well.

**Comment by danking** [comment]
@daniel-goldstein are you owning review for this?

**Comment by daniel-goldstein** [comment]
Apologies, I'll review this today

File: batch/sql/estimated-current.sql
Comment (daniel-goldstein) [inline batch/sql/estimated-current.sql:321]: missing an added column to the jobs table.

**Review by daniel-goldstein — COMMENTED** [review]
I think this looks good, but would appreciate a final review by Dan especially on the new procedure bit.

**Comment by jigold** [comment]
@danking Can you please take a look?

**Comment by danking** [comment]
I'm a bit flat out this week but will endeavor to look before Friday.

File: batch/sql/estimated-current.sql
Comment (danking) [inline batch/sql/estimated-current.sql:313]: OK, so this is a really key and important aspect: `batch_id, job_id` is still the unique identifier of a job. Groups are merely some organizational structure atop all this. I think verbiage to this effect should make it into the RFC. What's the unique identifier for the relevant concepts. I think its:
1. Job: `batch_id, job_id`
2. Group: `batch_id, job_group_id`.
3. Batch: `batch_id`

File: batch/sql/estimated-current.sql
Comment (danking) [inline batch/sql/estimated-current.sql:225]: Do I understand correctly that, in this PR, we are setting up database structures for tracking states, cancelling, and billing but *we are not adding* structure of creation of new groups. Instead, in this PR, every batch has just one group (so we can piggy back on batch creation).

File: batch/sql/estimated-current.sql
Comment (danking) [reply batch/sql/estimated-current.sql:322]: OK, so, a job is the member of exactly one group. Groups may or may not have extra structure, but every job is member of exactly one group.

File: batch/sql/estimated-current.sql
Comment (danking) [inline batch/sql/estimated-current.sql:326]: Down in the unchanged code, `job_parents` indicates that jobs can have parents in different groups. Verify I understand correctly:

> This is fine because, by design, a job can be cancelled while its children are still Pending. The cancellation loop will eventually notice the cancelled parent and mark the children as cancelled. A cancelled job is only concerned with its member jobs (the jobs with its id as their job_group_id).

File: batch/sql/estimated-current.sql
Comment (danking) [inline batch/sql/estimated-current.sql:395]: OK, again, confirm for me:

> This is now a poorly named table: it's really `job_group_attributes`. A subsequent PR will rename this table.

And also confirm for me:

> The attributes on the root group id are displayed in the UI as the batch's attributes.

File: batch/sql/estimated-current.sql
Comment (danking) [inline batch/sql/estimated-current.sql:438]: Note to ourselves, these tables should be dropped entirely soon anyway: https://github.com/hail-is/hail/issues/13656

File: batch/sql/estimated-current.sql
Comment (danking) [inline batch/sql/estimated-current.sql:705]: I'm a bit surprised by this. Does this substantially simplify other code? Having a self-loop in the parents table is confusing because it violates the statement

> if `(_, y, z, _)` is in `job_groups_parents` then `z` is the parent of `y`.

File: batch/sql/estimated-current.sql
Comment (danking) [inline batch/sql/estimated-current.sql:1186]: Why the inner join? It has to be `expected_n_jobs`, right? This seems like natural code to be added by the subsequent PR that actually handles parallel creation of job groups.

File: batch/sql/estimated-current.sql
Comment (danking) [reply batch/sql/estimated-current.sql:225]: (I made this note here because batch_updates has no job_group_id).

File: batch/sql/estimated-current.sql
Comment (danking) [inline batch/sql/estimated-current.sql:1574]: This is pretty heady stuff! Worth a link to https://dev.mysql.com/doc/refman/8.0/en/cursors.html

File: batch/sql/estimated-current.sql
Comment (danking) [inline batch/sql/estimated-current.sql:1703]: Just to be clear: this has absolutely no effect in this PR because it is not possible to create job groups (other than the implicitly created root one)?

**Review by danking — DISMISSED** [review]
Mostly questions

File: batch/sql/estimated-current.sql
Comment (jigold) [reply batch/sql/estimated-current.sql:225]: Yes, you understand correctly. This is just to get job groups in the database and mirroring the correct state for the root job group and the batch.

`batch_updates` should never have a job group id as an update is a separate concept and an update can contain multiple job groups. The update is just the staged "transaction" of changes to be made to the batch rather than the job organization.

File: batch/sql/estimated-current.sql
Comment (jigold) [reply batch/sql/estimated-current.sql:326]: Not sure who wrote that statement whether it was you or me. That is correct except the last statement should be 

> A cancelled job is only concerned with its child jobs (the jobs with its id as their parent_id). 

Job groups have nothing to do with individual job level cancellation. They are only necessary for a bulk cancellation operation in which case any cancelled jobs will propogate that cancellation event down to their children.

File: batch/sql/estimated-current.sql
Comment (jigold) [reply batch/sql/estimated-current.sql:395]: Agreed the table name sucks (all of the new tables that have been updated have bad names). Figuring out how to do the rename online will require trying to write to both the old and new table name. I don't think we can just wait for the container to redeploy and I don't want to copy the tables completely. Writing to both the old and new doesn't seem so bad if we build some infrastructure for this. I just haven't thought about it deeply yet.

Nothing has changed about the UI. Since the job group id is always "0" and is not a primary key (yet), it's just as if there's a dummy extra column right now. Shouldn't change any existing queries.

File: batch/sql/estimated-current.sql
Comment (jigold) [reply batch/sql/estimated-current.sql:705]: I believe if we don't do this, then we'll need to split the queries into two in a bunch of places or maybe use loops. Once to do the operation on the group of interest and then another to do it on all its parents. Here's an example of billing propagation up the tree and then cancellation down the tree:

```sql
    INSERT INTO aggregated_batch_resources_v3 (batch_id, job_group_id, resource_id, token, `usage`)
    SELECT attempt_resources.batch_id,
      job_group_parents.parent_id,
      attempt_resources.deduped_resource_id,
      rand_token,
      msec_diff_rollup * quantity
    FROM attempt_resources
    LEFT JOIN jobs ON attempt_resources.batch_id = jobs.batch_id AND attempt_resources.job_id = jobs.job_id
    LEFT JOIN job_group_parents ON jobs.batch_id = job_group_parents.batch_id AND jobs.job_group_id = job_group_parents.job_group_id
    JOIN aggregated_batch_resources_v2 ON
      aggregated_batch_resources_v2.batch_id = attempt_resources.batch_id AND
      aggregated_batch_resources_v2.resource_id = attempt_resources.resource_id AND
      aggregated_batch_resources_v2.token = rand_token
    WHERE attempt_resources.batch_id = NEW.batch_id AND attempt_resources.job_id = NEW.job_id AND attempt_resources.attempt_id = NEW.attempt_id AND migrated = 1
    ON DUPLICATE KEY UPDATE `usage` = aggregated_batch_resources_v3.`usage` + msec_diff_rollup * quantity;
```

```sql
  INSERT INTO batch_inst_coll_cancellable_resources (batch_id, update_id, job_group_id, inst_coll, token,
    n_ready_cancellable_jobs, ready_cancellable_cores_mcpu, n_creating_cancellable_jobs, n_running_cancellable_jobs,
    running_cancellable_cores_mcpu)
  SELECT batch_inst_coll_cancellable_resources.batch_id, batch_inst_coll_cancellable_resources.update_id, job_group_parents.parent_id, inst_coll, 0,
    -1 * (@n_ready_cancellable_jobs := COALESCE(SUM(n_ready_cancellable_jobs), 0)),
    -1 * (@ready_cancellable_cores_mcpu := COALESCE(SUM(ready_cancellable_cores_mcpu), 0)),
    -1 * (@n_creating_cancellable_jobs := COALESCE(SUM(n_creating_cancellable_jobs), 0)),
    -1 * (@n_running_cancellable_jobs := COALESCE(SUM(n_running_cancellable_jobs), 0)),
    -1 * (@running_cancellable_cores_mcpu := COALESCE(SUM(running_cancellable_cores_mcpu), 0))
  FROM batch_inst_coll_cancellable_resources
  INNER JOIN batch_updates ON batch_inst_coll_cancellable_resources.batch_id = batch_updates.batch_id AND
    batch_inst_coll_cancellable_resources.update_id = batch_updates.update_id
  INNER JOIN job_group_parents ON batch_inst_coll_cancellable_resources.batch_id = job_group_parents.batch_id AND
    batch_inst_coll_cancellable_resources.job_group_id = job_group_parents.job_group_id
  WHERE job_group_parents.batch_id = in_batch_id AND
    job_group_parents.job_group_id = in_job_group_id AND
    batch_updates.committed
  GROUP BY batch_inst_coll_cancellable_resources.batch_id, batch_inst_coll_cancellable_resources.update_id,
    job_group_parents.parent_id, inst_coll
  ON DUPLICATE KEY UPDATE
    n_ready_cancellable_jobs = n_ready_cancellable_jobs - @n_ready_cancellable_jobs,
    ready_cancellable_cores_mcpu = ready_cancellable_cores_mcpu - @ready_cancellable_cores_mcpu,
    n_creating_cancellable_jobs = n_creating_cancellable_jobs - @n_creating_cancellable_jobs,
    n_running_cancellable_jobs = n_running_cancellable_jobs - @n_running_cancellable_jobs,
    running_cancellable_cores_mcpu = running_cancellable_cores_mcpu - @running_cancellable_cores_mcpu;

  # delete all rows that are children of this job group
  DELETE batch_inst_coll_cancellable_resources FROM batch_inst_coll_cancellable_resources
  LEFT JOIN batch_updates ON batch_inst_coll_cancellable_resources.batch_id = batch_updates.batch_id AND
    batch_inst_coll_cancellable_resources.update_id = batch_updates.update_id
  INNER JOIN job_group_parents ON batch_inst_coll_cancellable_resources.batch_id = job_group_parents.batch_id AND
    batch_inst_coll_cancellable_resources.job_group_id = job_group_parents.job_group_id
  WHERE batch_inst_coll_cancellable_resources.batch_id = in_batch_id AND
    job_group_parents.parent_id = in_job_group_id AND
    batch_updates.committed;
```

Comment (jigold) [reply batch/sql/estimated-current.sql:705]: There might be other queries in the client code that having the self loop is helpful.

File: batch/sql/estimated-current.sql
Comment (jigold) [reply batch/sql/estimated-current.sql:1186]: I don't remember why I used an inner join. I don't think it really matters. No, you need to add this query here. The whole point of this PR is to make sure the batches table and job groups table are in sync. If the batch is updated, then the corresponding job group's state must be updated.

File: batch/sql/estimated-current.sql
Comment (jigold) [reply batch/sql/estimated-current.sql:1703]: I believe that is correct. I believe I can move this into a different PR that will need to be done after the job groups tables are fully populated that allow for job_group_id != 0 as well as having explicit primary keys with job_group_id in the tables for `aggregated_batch_resources`, `batches_cancelled` etc.

**Comment by jigold** [comment]
@danking I'm keeping the changes requested state on until I get the full scope of changes that need to be made. Would be good to get your feedback first after my responses so I can do all changes in one pass.

**Comment by jigold** [comment]
@danking Would it be possible to get another review by the end of this week? Not urgent, but would like to keep making progress on this.

File: batch/sql/estimated-current.sql
Comment (jigold) [reply batch/sql/estimated-current.sql:705]: This is I believe the major outstanding point of discussion.

Comment (danking) [reply batch/sql/estimated-current.sql:705]: Ah, I think this table has the wrong name. Knowing only the name, I thought this was true:

> if `(_, y, z, _)` is in `job_groups_parents` then `z` is the parent of `y`.

But after seeing the self-loop I thought this was true:

> if `(_, y, z, _)` is in `job_groups_parents` then `z` is the parent of `y` or `z` == `y`.

But now I see that what this *actually* is is something like `job_groups_needing_to_be_updated_when_a_job_in_this_group_changes`. It's not the ancestors (because of the self loop) and it's not the parents (because it includes ancestors and a self loop).

We need a clear and correct name for this table, but I do not object to the table itself.

File: batch/sql/estimated-current.sql
Comment (danking) [reply batch/sql/estimated-current.sql:395]: Just take batch offline long enough to do the renames. Renames should be near instantaneous. We have no SLAs and even if we did we wouldn't have 100% uptime! 😄

File: batch/sql/estimated-current.sql
Comment (danking) [reply batch/sql/estimated-current.sql:326]: er, sorry I meant to say:
> A cancelled job group is only concerned with its member jobs (the jobs with its id as their job_group_id).

And maybe I should have added:
> A cancelled job group need not care about the descendants of its jobs. The job cancelling loop will handle them.

File: batch/sql/estimated-current.sql
Comment (danking) [reply batch/sql/estimated-current.sql:1574]: For this I meant we need a comment with that link

**Review by danking — DISMISSED** [review]
responded to comments

File: batch/sql/estimated-current.sql
Comment (jigold) [reply batch/sql/estimated-current.sql:395]: Is it okay if we do the renames in a separate PR? I don't want to do too much in one set of changes and there has to be a separate follow up PR to change the foreign key constraints and primary keys as well. We'll also want to rename the indices on those tables.

File: batch/sql/estimated-current.sql
Comment (jigold) [reply batch/sql/estimated-current.sql:705]: How about `job_group_tree`?

**Comment by jigold** [comment]
@danking I changed the name `job_group_parents` => `job_group_tree`.

**Comment by jigold** [comment]
Can I assume this is basically done so I can move on to the next part of the stack?

**Comment by danking** [comment]
I still dislike job_group_tree because trees are usually represented in terms of their ancestor or parent-child relationships, neither of which are what this is. That said, I don't think we should block this PR on a naming quibble.

We can do renames separately. We should update the RFC to be clear about the unique identifiers of the three things groups, jobs, and batches. (In particular, jobs are uniquely identifier by batch id and job id, group id is not part of it).

**Comment by danking** [comment]
In Order Theory, this operation is called "the upper/downward closure of x" https://en.wikipedia.org/wiki/Upper_set . `groups_affected_by` maybe? `group_self_and_ancestors`?

**Comment by jigold** [comment]
How about `job_group_directed_sets` or `job_group_lattice`? https://en.wikipedia.org/wiki/Directed_set https://en.wikipedia.org/wiki/Lattice_(order)

**Comment by danking** [comment]
My issue with both of those names is the same as with `_tree`. It's not clear if you're storing the non-transitive or the transitive relation & its not clear if self-edges are included. I want a name that unambiguously says "the self-edge and all the ancestor edges are in here" or a name that is more domain-specific like "job groups that need to be updated when a job in this job group is changed". 

`job_group_self_and_ancestor` feels like the shortest name so far that satisfies my concerns, but I'm open to other ones that are clear about self-edge-ness and transitive-ness.

File: batch/sql/estimated-current.sql
Comment (jigold) [reply batch/sql/estimated-current.sql:396]: See #13810 for table renames

**Comment by jigold** [comment]
Ok. It's been renamed.

**Review by danking — APPROVED** [review]
I'm currently assigned, so I'll approve, but ping to @daniel-goldstein if you had remaining concerns.

**Comment by daniel-goldstein** [comment]
I think this thread addressed everything, looks good to me!

**Comment by jigold** [comment]
@daniel-goldstein Can you please check the last commit real quick? Should just be a simple rename, but want someone to double check I didn't do something bad.

File: batch/sql/setup-job-groups.sql
Comment (daniel-goldstein) [inline batch/sql/setup-job-groups.sql:402]: Good rename but I'm seeing multiple occurrences of `parent_id` and `job_parents` in the migration itself. Feels like any old names in here should error in CI so I'm not worried once this is green.

Comment (daniel-goldstein) [reply batch/sql/setup-job-groups.sql:402]: Oh yes oops, I fully read that as `job_group_parents`.

Comment (daniel-goldstein) [reply batch/sql/setup-job-groups.sql:402]: Great rename, it'll be much easier to keep these column names straight now that there aren't two columns that mean different things across different tables

**Pull request merged**
